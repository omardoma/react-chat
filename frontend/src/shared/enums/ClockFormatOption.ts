export enum ClockFormatOption {
  TwelveHours = 'hh',
  TwentyFourHours = 'HH'
}
