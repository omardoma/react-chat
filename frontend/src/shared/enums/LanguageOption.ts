export enum LanguageOption {
  English = 'en',
  German = 'de'
}
