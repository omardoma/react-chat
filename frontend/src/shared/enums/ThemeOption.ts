export enum ThemeOption {
  Light = 'light',
  Dark = 'dark'
}
