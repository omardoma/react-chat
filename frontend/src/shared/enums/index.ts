export * from './ClockFormatOption';
export * from './LanguageOption';
export * from './ThemeOption';
