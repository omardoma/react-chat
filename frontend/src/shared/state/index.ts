export { useGlobalState, GlobalStateProvider } from './context';
