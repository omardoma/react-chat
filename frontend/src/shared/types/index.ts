export * from './Message';
export * from './Settings';
