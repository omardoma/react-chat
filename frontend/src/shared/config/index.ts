import { ClockFormatOption, ThemeOption, LanguageOption } from '../enums';
import { getUserLanguage, generateId } from '../utils';

const userLanguage = getUserLanguage() as LanguageOption;

const defaultLanguage = Object.values(LanguageOption).includes(userLanguage)
  ? userLanguage
  : LanguageOption.English;

export const DEFAULT_SETTINGS = {
  username: `guest${generateId()}`,
  theme: ThemeOption.Light,
  clockFormat: ClockFormatOption.TwelveHours,
  sendOnCtrlEnter: false,
  language: defaultLanguage
};

export const SOCKET_SERVER_URL = 'http://localhost:3000';
