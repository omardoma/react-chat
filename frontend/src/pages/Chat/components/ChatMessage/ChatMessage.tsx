import React from 'react';

import { makeStyles, Theme } from '@material-ui/core/styles';
import { FormattedDate } from 'react-intl';
import { isSameDay } from 'date-fns';

import { Message } from '../../../../shared/types';
import { ClockFormatOption } from '../../../../shared/enums';

const useStyles = makeStyles((theme: Theme) => ({
  timestamp: {
    paddingLeft: '20px',
    paddingRight: '20px',
    alignSelf: ({ isSent }: ChatMessageProps) =>
      isSent ? 'flex-end' : 'flex-start',
    marginTop: 0,
    marginBottom: '0.25rem',
    color: theme.palette.text.hint
  },
  bubble: {
    position: 'relative',
    maxWidth: '50%',
    wordBreak: 'break-word',
    whiteSpace: 'pre-wrap',
    marginBottom: '0.5rem',
    borderRadius: '0.5rem',
    padding: '0.5rem',
    alignSelf: ({ isSent }: ChatMessageProps) =>
      isSent ? 'flex-end' : 'flex-start',
    backgroundColor: ({ isSent }: ChatMessageProps) =>
      isSent ? theme.palette.primary.main : theme.palette.grey[400],
    '&.sent::after': {
      content: '""',
      position: 'absolute',
      top: 0,
      right: '10px',
      width: 0,
      height: 0,
      border: '10px solid transparent',
      borderBottomColor: theme.palette.primary.main,
      borderTop: 0,
      borderRight: 0,
      marginLeft: '-5px',
      marginTop: '-10px'
    },
    '&.received::after': {
      content: '""',
      position: 'absolute',
      top: 0,
      left: '10px',
      width: 0,
      height: 0,
      border: '10px solid transparent',
      borderBottomColor: theme.palette.grey[400],
      borderTop: 0,
      borderLeft: 0,
      marginRight: '-5px',
      marginTop: '-10px'
    },
    '& p': {
      margin: 0
    }
  }
}));

interface ChatMessageProps {
  message: Message;
  isSent: boolean;
  clockFormat: ClockFormatOption;
}

function ChatMessage(props: ChatMessageProps): JSX.Element {
  const classes = useStyles(props);
  const { message, isSent, clockFormat } = props;
  const timestamp = isSameDay ? (
    <FormattedDate
      value={message.createdAt}
      hour="numeric"
      minute="numeric"
      hour12={clockFormat === ClockFormatOption.TwelveHours}
    />
  ) : (
    <FormattedDate
      value={message.createdAt}
      day="numeric"
      month="numeric"
      year="numeric"
      hour="numeric"
      minute="numeric"
      hour12={clockFormat === ClockFormatOption.TwelveHours}
    />
  );
  return (
    <>
      <p className={classes.timestamp}>
        {!isSent && <span>{message.senderUsername}, </span>}
        {timestamp}
      </p>
      <div className={`${classes.bubble} ${isSent ? 'sent' : 'received'}`}>
        <p>{message.content}</p>
      </div>
    </>
  );
}

export default ChatMessage;
