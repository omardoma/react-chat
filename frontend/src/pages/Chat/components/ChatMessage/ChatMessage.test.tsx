import React from 'react';

import { render } from 'test-utils';

import { Message } from '../../../../shared/types';
import { ClockFormatOption } from '../../../../shared/enums';

import ChatMessage from './ChatMessage';

describe('ChatMessage', () => {
  it('renders a sent ChatMessage component when isSent is true with 12 Hours clock format', () => {
    // Arrange
    const message: Message = {
      id: '1',
      content: 'Hey There',
      senderUsername: 'guest1',
      createdAt: new Date(2020, 6, 25).toISOString()
    };
    const isSent = true;
    const clockFormat = ClockFormatOption.TwelveHours;
    const { asFragment } = render(
      <ChatMessage
        message={message}
        isSent={isSent}
        clockFormat={clockFormat}
      />
    );
    // Assert
    expect(asFragment()).toMatchSnapshot();
  });

  it('renders a sent ChatMessage component when isSent is true with 24 Hours clock format', () => {
    // Arrange
    const message: Message = {
      id: '1',
      content: 'Hey There',
      senderUsername: 'guest1',
      createdAt: new Date(2020, 6, 25).toISOString()
    };
    const isSent = true;
    const clockFormat = ClockFormatOption.TwentyFourHours;
    const { asFragment } = render(
      <ChatMessage
        message={message}
        isSent={isSent}
        clockFormat={clockFormat}
      />
    );
    // Assert
    expect(asFragment()).toMatchSnapshot();
  });

  it('renders a received ChatMessage component when isSent is false with 12 Hours clock format', () => {
    // Arrange
    const message: Message = {
      id: '1',
      content: 'Hey There',
      senderUsername: 'guest1',
      createdAt: new Date(2020, 6, 25).toISOString()
    };
    const isSent = false;
    const clockFormat = ClockFormatOption.TwelveHours;
    const { asFragment } = render(
      <ChatMessage
        message={message}
        isSent={isSent}
        clockFormat={clockFormat}
      />
    );
    // Assert
    expect(asFragment()).toMatchSnapshot();
  });

  it('renders a received ChatMessage component when isSent is false with 24 Hours clock format', () => {
    // Arrange
    const message: Message = {
      id: '1',
      content: 'Hey There',
      senderUsername: 'guest1',
      createdAt: new Date(2020, 6, 25).toISOString()
    };
    const isSent = false;
    const clockFormat = ClockFormatOption.TwentyFourHours;
    const { asFragment } = render(
      <ChatMessage
        message={message}
        isSent={isSent}
        clockFormat={clockFormat}
      />
    );
    // Assert
    expect(asFragment()).toMatchSnapshot();
  });
});
