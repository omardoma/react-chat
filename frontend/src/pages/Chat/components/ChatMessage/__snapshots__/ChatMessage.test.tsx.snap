// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ChatMessage renders a received ChatMessage component when isSent is false with 12 Hours clock format 1`] = `
<DocumentFragment>
  <p
    class="makeStyles-timestamp-9 makeStyles-timestamp-11"
  >
    <span>
      guest1, 
    </span>
    12:00 AM
  </p>
  <div
    class="makeStyles-bubble-10 makeStyles-bubble-12 received"
  >
    <p>
      Hey There
    </p>
  </div>
</DocumentFragment>
`;

exports[`ChatMessage renders a received ChatMessage component when isSent is false with 24 Hours clock format 1`] = `
<DocumentFragment>
  <p
    class="makeStyles-timestamp-13 makeStyles-timestamp-15"
  >
    <span>
      guest1, 
    </span>
    00:00
  </p>
  <div
    class="makeStyles-bubble-14 makeStyles-bubble-16 received"
  >
    <p>
      Hey There
    </p>
  </div>
</DocumentFragment>
`;

exports[`ChatMessage renders a sent ChatMessage component when isSent is true with 12 Hours clock format 1`] = `
<DocumentFragment>
  <p
    class="makeStyles-timestamp-1 makeStyles-timestamp-3"
  >
    12:00 AM
  </p>
  <div
    class="makeStyles-bubble-2 makeStyles-bubble-4 sent"
  >
    <p>
      Hey There
    </p>
  </div>
</DocumentFragment>
`;

exports[`ChatMessage renders a sent ChatMessage component when isSent is true with 24 Hours clock format 1`] = `
<DocumentFragment>
  <p
    class="makeStyles-timestamp-5 makeStyles-timestamp-7"
  >
    00:00
  </p>
  <div
    class="makeStyles-bubble-6 makeStyles-bubble-8 sent"
  >
    <p>
      Hey There
    </p>
  </div>
</DocumentFragment>
`;
